### DISCUSSION TITLE
Clarify on the technical implementation of Batect's caches

### DISCUSSION THREAD

**Discussion by sschuberth** [batect/batect#1009] [General]
While reading though [[1]] and [[2]] I was wondering whether Batect's cache "simply" is a wrapper around BuildKits's [--mount=type=cache](https://github.com/moby/buildkit/blob/master/frontend/dockerfile/docs/syntax.md#run---mounttypecache), or something homegrown. Could you maybe elaborate?

[1]: https://batect.dev/docs/concepts/caches/
[2]: https://batect.dev/docs/reference/config/containers#caches

**Comment by charleskorn** [comment]
Hi @sschuberth, they're similar ideas, but used at different times:

* BuildKit's `--mount=type=cache` (and options like `--cache-from` and `--cache-to`) are used when building images. `--cache-from` and `--cache-to` help cache layers of an image, speeding up the image build, `--mount=type=cache` allows you to mount a directory into the build environment to persist files such as downloaded package files between image builds without storing them in the image.

  Batect doesn't (yet) have support for these.

* Batect's caches are used when running containers, and are mounted into the running container. This allows you to mount a directory into the container to persist files between tasks.

**Reply by sschuberth** [reply]
Thanks for the clarification. I keep forgetting that Batect actually never builds images itself, also not implicitly. So put simply, `--mount=type=cache` is a build time cache, whereas `- type: cache` is a runtime cache. Though both seem to be backed by a read / write layer (as opposed to a volume / bind mount). Correct?

**Comment by charleskorn** [comment]
> So put simply, --mount=type=cache is a build time cache, whereas - type: cache is a runtime cache.

Yep.

> Though both seem to be backed by a read / write layer (as opposed to a volume / bind mount).

I'm not sure of the exact mechanism used by BuildKit. For Batect's runtime caches, they are either volume or bind mounts. The default is a volume mount, and you can use a bind mount by passing [`--cache-type=directory`](https://batect.dev/docs/reference/cli/#--cache-type) on the command line.
